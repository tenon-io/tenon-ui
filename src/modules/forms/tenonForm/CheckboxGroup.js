import React, { forwardRef } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import FocusCatcher from '../../utils/components/FocusCatcher';
import FeedbackBlock from './FeedbackBlock';

/**
 * @component
 * A default view component to use with the tenon-ui
 * smart checkbox group controller.
 *
 * @prop required {string} legend - Text to show as the legend.
 * @prop required {object} options - Key mapping object to show
 *                the checkboxes from. The keys are used as
 *                radiobutton values, the values as the labels.
 * @prop {string} contentHintText - Text to show as content
 *                hint of the checkbox group.
 * @prop {string} errorText - Text to show as error message
 *                of the checkbox group.
 * @prop required {function} getLabelProps - Prop getter function
 *              from the smart checkbox group controller for each <label>
 *              of the group. The autoIdPostfix config option should
 *              be set with the name of the checkbox input or the id
 *              linkage will fail.
 * @prop required {function} getCheckboxProps - Prop getter function
 *              from the smart checkbox group controller for each <input>
 *              of the group. It is required to set a name here or the
 *              component will fail.
 * @prop required {function} getFieldsetProps - Prop getter function
 *              from the smart checkbox group controller for the fieldset
 *              element.
 * @prop {boolean} showError - A flag to indicate when the
 *              error container should be displayed.
 * @prop {string} required - Set to 'required' to mark the
 *              checkbox group container as required.
 * @prop {string} className - An optional class string to transfer to the
 *              className prop of the select element.
 * @prop {string} requiredText - And optional string value to display
 *              after the legend indicating a required situation.
 *@prop {boolean} required - An optional boolean value indicating whether
 *              the current view component is attached to a required
 *              data model.
 */
const CheckboxGroup = forwardRef(
    (
        {
            legend,
            options,
            getCheckboxProps,
            getLabelProps,
            contentHintText,
            errorText,
            showError,
            required,
            requiredText,
            className
        },
        ref
    ) => (
        <fieldset className={classNames('form-group', className)}>
            <legend>
                {legend}
                {required ? (
                    <span className="required">
                        &nbsp;
                        {requiredText || '*'}
                    </span>
                ) : null}
                {contentHintText && (
                    <span className="visually-hidden">
                        &nbsp;
                        {contentHintText}
                    </span>
                )}
                {errorText && showError && (
                    <span className="visually-hidden">
                        &nbsp;
                        {errorText}
                    </span>
                )}
            </legend>
            <FocusCatcher>
                <ul className="checkbox-container">
                    {Object.keys(options).map((option, i) => (
                        <li key={option}>
                            <div className="checkbox-wrapper">
                                <input
                                    {...getCheckboxProps({
                                        name: option,
                                        focusElement: i === 0
                                    })}
                                    ref={i === 0 ? ref : null}
                                />
                                <label
                                    {...getLabelProps({
                                        autoIdPostfix: i === 0 ? '' : option
                                    })}
                                >
                                    {options[option]}
                                </label>
                            </div>
                        </li>
                    ))}
                </ul>
            </FocusCatcher>
            <FeedbackBlock
                errorText={errorText}
                contentHintText={contentHintText}
                showError={showError}
                aria-hidden="true"
            />
        </fieldset>
    )
);

CheckboxGroup.displayName = 'CheckboxGroup';

CheckboxGroup.propTypes = {
    legend: PropTypes.string.isRequired,
    options: PropTypes.object.isRequired,
    getCheckboxProps: PropTypes.func.isRequired,
    getLabelProps: PropTypes.func.isRequired,
    required: PropTypes.bool,
    contentHintText: PropTypes.string,
    errorText: PropTypes.string,
    showError: PropTypes.bool,
    className: PropTypes.string,
    requiredText: PropTypes.string
};

export default CheckboxGroup;
