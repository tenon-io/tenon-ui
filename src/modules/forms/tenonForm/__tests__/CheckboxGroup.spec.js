jest.mock('../../../utils/components/FocusCatcher');

import React, { createRef } from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';
import CheckboxGroup from '../CheckboxGroup';

describe('CheckboxGroup', () => {
    afterEach(cleanup);

    it('should render a basic checkbox group', () => {
        const { container } = render(
            <CheckboxGroup
                legend="Test checkbox group"
                options={{
                    one: 'One',
                    two: 'Two'
                }}
                getLegendProps={() => ({ id: 'bar' })}
                getLabelProps={props => ({
                    htmlFor: `foo-${props.autoIdPostfix}`
                })}
                getCheckboxProps={({ name }) => ({
                    id: `foo-${name}`,
                    name,
                    onChange: jest.fn()
                })}
                getCheckboxGroupProps={() => ({
                    'aria-labelledby': 'bar',
                    role: 'group'
                })}
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render an error text when invalid', () => {
        const { container } = render(
            <CheckboxGroup
                legend="Test checkbox group"
                options={{
                    one: 'One',
                    two: 'Two'
                }}
                getLegendProps={() => ({ id: 'bar' })}
                getLabelProps={props => ({
                    htmlFor: `foo-${props.autoIdPostfix}`
                })}
                getCheckboxProps={({ name }) => ({
                    id: `foo-${name}`,
                    name,
                    onChange: jest.fn()
                })}
                getCheckboxGroupProps={() => ({
                    'aria-describedby': 'err',
                    'aria-labelledby': 'bar',
                    role: 'group'
                })}
                getErrorProps={() => ({ id: 'err' })}
                errorText="Test error"
                showError={true}
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render a content hint if given', () => {
        const { container } = render(
            <CheckboxGroup
                legend="Test checkbox group"
                options={{
                    one: 'One',
                    two: 'Two'
                }}
                getLegendProps={() => ({ id: 'bar' })}
                getLabelProps={props => ({
                    htmlFor: `foo-${props.autoIdPostfix}`
                })}
                getCheckboxProps={({ name }) => ({
                    id: `foo-${name}`,
                    name,
                    onChange: jest.fn()
                })}
                getCheckboxGroupProps={() => ({
                    'aria-describedby': 'inf',
                    'aria-labelledby': 'bar',
                    role: 'group'
                })}
                getContentHintProps={() => ({ id: 'inf' })}
                contentHintText="Content hint test"
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should allow easy focus of the internal radio group', () => {
        const checkboxGroupRef = createRef();

        const { container } = render(
            <CheckboxGroup
                ref={checkboxGroupRef}
                legend="Test radio group"
                options={{
                    one: 'One',
                    two: 'Two'
                }}
                getLegendProps={() => ({ id: 'bar' })}
                getLabelProps={props => ({
                    htmlFor: `foo-${props.autoIdPostfix}`
                })}
                getCheckboxProps={({ name }) => ({
                    id: `foo-${name}`,
                    name,
                    onChange: jest.fn()
                })}
                getCheckboxGroupProps={() => ({
                    'aria-labelledby': 'bar',
                    role: 'group'
                })}
            />
        );
        expect(container.querySelector('[role="group"]')).toEqual(
            checkboxGroupRef.current
        );
    });
});
