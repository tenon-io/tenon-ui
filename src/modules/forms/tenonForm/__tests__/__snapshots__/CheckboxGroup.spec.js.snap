// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup should render a basic checkbox group 1`] = `
<fieldset
  class="form-group"
>
  <legend>
    Test checkbox group
  </legend>
  <div>
    <span>
      FocusCatcher
    </span>
    <ul
      class="checkbox-container"
    >
      <li>
        <div
          class="checkbox-wrapper"
        >
          <input
            id="foo"
            name="one"
          />
          <label
            for="foo"
          >
            One
          </label>
        </div>
      </li>
      <li>
        <div
          class="checkbox-wrapper"
        >
          <input
            id="foo-two"
            name="two"
          />
          <label
            for="foo-two"
          >
            Two
          </label>
        </div>
      </li>
    </ul>
  </div>
</fieldset>
`;

exports[`CheckboxGroup should render a content hint if given 1`] = `
<fieldset
  class="form-group"
>
  <legend>
    Test checkbox group
    <span
      class="visually-hidden"
    >
       
      Content hint test
    </span>
  </legend>
  <div>
    <span>
      FocusCatcher
    </span>
    <ul
      class="checkbox-container"
    >
      <li>
        <div
          class="checkbox-wrapper"
        >
          <input
            id="foo"
            name="one"
          />
          <label
            for="foo"
          >
            One
          </label>
        </div>
      </li>
      <li>
        <div
          class="checkbox-wrapper"
        >
          <input
            id="foo-two"
            name="two"
          />
          <label
            for="foo-two"
          >
            Two
          </label>
        </div>
      </li>
    </ul>
  </div>
  <div
    aria-hidden="true"
    class="info-wrapper"
  >
    <span
      class="icon"
    />
    <span>
      Content hint test
    </span>
  </div>
</fieldset>
`;

exports[`CheckboxGroup should render an error text when invalid 1`] = `
<fieldset
  class="form-group"
>
  <legend>
    Test checkbox group
    <span
      class="visually-hidden"
    >
       
      Test error
    </span>
  </legend>
  <div>
    <span>
      FocusCatcher
    </span>
    <ul
      class="checkbox-container"
    >
      <li>
        <div
          class="checkbox-wrapper"
        >
          <input
            id="foo"
            name="one"
          />
          <label
            for="foo"
          >
            One
          </label>
        </div>
      </li>
      <li>
        <div
          class="checkbox-wrapper"
        >
          <input
            id="foo-two"
            name="two"
          />
          <label
            for="foo-two"
          >
            Two
          </label>
        </div>
      </li>
    </ul>
  </div>
  <div
    aria-hidden="true"
    class="error-wrapper"
  >
    <span
      class="icon"
    />
    <span>
      Test error
    </span>
  </div>
</fieldset>
`;
